File: backend/services/boost_service.py
"""Listing Boost service — pay with credits or INR, feed weighting, auto-expiry."""
from __future__ import annotations
import logging
from datetime import datetime, timezone, timedelta
from bson import ObjectId
from fastapi import HTTPException

from database import get_db
from services import wallet_service, phase4_service

logger = logging.getLogger(__name__)

# Cost table
BOOST_PLANS = {
    3:  {"credits": 300,  "paise": 9900},   # ₹99
    7:  {"credits": 600,  "paise": 19900},  # ₹199
    14: {"credits": 1000, "paise": 34900},  # ₹349
}


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _validate_plan(duration_days: int) -> dict:
    plan = BOOST_PLANS.get(int(duration_days))
    if not plan:
        raise HTTPException(400, f"Invalid duration_days. Allowed: {list(BOOST_PLANS)}")
    return plan


async def _get_owned_listing(listing_id: str, vendor_id: str) -> dict:
    db = get_db()
    if not ObjectId.is_valid(listing_id):
        raise HTTPException(400, "Invalid listing id")
    listing = await db.listings.find_one({"_id": ObjectId(listing_id), "is_deleted": {"$ne": True}})
    if not listing:
        raise HTTPException(404, "Listing not found")
    if str(listing.get("vendor_id")) != vendor_id:
        raise HTTPException(403, "Only the listing owner can boost")
    if listing.get("is_takendown"):
        raise HTTPException(403, "Listing is under admin review")
    return listing


async def _apply_boost(listing_id: str, duration_days: int) -> dict:
    """Set boost_expires_at on the listing. Extends if already boosted."""
    db = get_db()
    now = datetime.now(timezone.utc)
    listing = await db.listings.find_one({"_id": ObjectId(listing_id)})
    current_expiry_str = listing.get("boost_expires_at")
    base_from = now
    if current_expiry_str:
        try:
            current_expiry = datetime.fromisoformat(str(current_expiry_str).replace("Z", "+00:00"))
            if current_expiry > now:
                base_from = current_expiry
        except Exception:  # noqa: BLE001
            pass
    new_expiry = base_from + timedelta(days=duration_days)
    await db.listings.update_one(
        {"_id": ObjectId(listing_id)},
        {"$set": {
            "boost_expires_at": new_expiry.isoformat(),
            "boost_duration_days": int(duration_days),
            "boost_activated_at": _now_iso(),
            "updated_at": _now_iso(),
        }},
    )
    updated = await db.listings.find_one({"_id": ObjectId(listing_id)})
    return {
        "listing_id": listing_id,
        "boost_expires_at": updated["boost_expires_at"],
        "boost_duration_days": updated["boost_duration_days"],
        "active": True,
    }


async def boost_with_credits(vendor_id: str, listing_id: str, duration_days: int) -> dict:
    plan = _validate_plan(duration_days)
    listing = await _get_owned_listing(listing_id, vendor_id)
    await wallet_service.spend_credits(
        vendor_id, plan["credits"],
        f"Boost listing {duration_days}d", "boost_listing", listing_id,
    )
    boost = await _apply_boost(listing_id, duration_days)
    # Notify vendor of activation
    from services import notification_service
    await notification_service.create(
        user_id=vendor_id, type_="boost",
        title="Listing boosted!",
        body=f"'{listing.get('title')}' is boosted for {duration_days} days.",
        action_url=f"/listing/{listing.get('slug')}",
    )
    return {"listing_id": listing_id, "boost": boost, "payment_method": "credits"}


async def boost_with_inr(vendor_id: str, listing_id: str, duration_days: int) -> dict:
    """Create a payment order. On simulate-success (or webhook), boost is activated."""
    plan = _validate_plan(duration_days)
    listing = await _get_owned_listing(listing_id, vendor_id)
    order = await phase4_service.create_payment_order(
        user_id=vendor_id,
        purpose="listing_boost",
        amount_paise=plan["paise"],
        ref_id=f"{listing_id}:{duration_days}",
    )
    return {
        "listing_id": listing_id,
        "payment": order,
        "payment_method": "inr",
        "duration_days": duration_days,
    }


async def activate_boost_from_payment(payment: dict) -> dict | None:
    """Called after successful payment for a listing_boost purpose."""
    ref = payment.get("ref_id") or ""
    if ":" not in ref:
        return None
    listing_id, dur = ref.split(":", 1)
    try:
        duration_days = int(dur)
    except ValueError:
        return None
    if not ObjectId.is_valid(listing_id):
        return None
    boost = await _apply_boost(listing_id, duration_days)
    from services import notification_service
    await notification_service.create(
        user_id=str(payment["user_id"]), type_="boost",
        title="Boost activated",
        body=f"Your listing is now boosted for {duration_days} days.",
        action_url="/vendor/dashboard",
    )
    return boost


async def list_my_boosted(vendor_id: str) -> list[dict]:
    db = get_db()
    now_iso = _now_iso()
    docs = await db.listings.find({
        "vendor_id": vendor_id, "is_deleted": {"$ne": True},
        "boost_expires_at": {"$gt": now_iso},
    }).sort([("boost_expires_at", 1)]).to_list(50)
    from services.listing_service import _serialize as _ls
    return [_ls(d) for d in docs]


async def expire_boosts_once() -> int:
    """Sweep expired boosts. Clear boost_expires_at & boost_duration_days when past."""
    db = get_db()
    now_iso = _now_iso()
    res = await db.listings.update_many(
        {"boost_expires_at": {"$lte": now_iso, "$ne": None}},
        {"$set": {"boost_expires_at": None, "boost_duration_days": None, "updated_at": now_iso}},
    )
    return res.modified_count


async def expire_boosts_loop(interval_seconds: int = 900) -> None:
    """Every 15 min: clear expired boosts."""
    import asyncio
    while True:
        try:
            n = await expire_boosts_once()
            if n:
                logger.info("Expired %d boost(s)", n)
        except Exception as e:  # noqa: BLE001
            logger.warning("boost expire loop err: %s", e)
        await asyncio.sleep(interval_seconds)

File: backend/src/services/boost.service.js
const Listing = require('../models/Listing');
const Reel = require('../models/Reel');
const { AppSettings, BoostPlan } = require('../models/Admin');
const walletService = require('./wallet.service');
const notificationService = require('./notification.service');
const ApiError = require('../utils/ApiError');
const logger = require('../utils/logger');
const { serializeListing } = require('./listing.service');

const validatePlan = async (durationDays) => {
  const days = parseInt(durationDays, 10);
  if (isNaN(days) || days <= 0) {
    throw ApiError.badRequest('Invalid duration days.');
  }

  const plan = await BoostPlan.findOne({ duration_days: days, is_active: true, is_deleted: { $ne: true } });
  if (!plan) {
    throw ApiError.badRequest(`No active listing boost plan found for ${days} days.`);
  }

  return {
    credits: plan.credits_cost,
    paise: Math.round(plan.price_inr * 100),
  };
};

const getOwnedListing = async (listingId, vendorId) => {
  const listing = await Listing.findOne({ _id: listingId, is_deleted: { $ne: true } });
  if (!listing) {
    throw ApiError.notFound('Listing not found');
  }
  if (listing.vendor_id !== vendorId) {
    throw ApiError.forbidden('Only the listing owner can boost');
  }
  if (listing.is_takendown) {
    throw ApiError.forbidden('Listing is under admin review');
  }
  return listing;
};

const applyBoost = async (listingId, durationDays) => {
  const now = new Date();
  const listing = await Listing.findById(listingId);
  if (!listing) {
    throw ApiError.notFound('Listing not found');
  }

  const currentExpiryStr = listing.boost_expires_at;
  let baseFrom = now;
  if (currentExpiryStr) {
    try {
      const currentExpiry = new Date(currentExpiryStr);
      if (currentExpiry > now) {
        baseFrom = currentExpiry;
      }
    } catch {}
  }

  const newExpiry = new Date(baseFrom.getTime() + durationDays * 24 * 60 * 60 * 1000);
  const nowIso = new Date().toISOString();
  await Listing.updateOne(
    { _id: listingId },
    {
      $set: {
        boost_expires_at: newExpiry.toISOString(),
        boost_duration_days: parseInt(durationDays, 10),
        boost_activated_at: nowIso,
        updated_at: nowIso,
      },
    }
  );

  const updated = await Listing.findById(listingId);
  return {
    listing_id: listingId,
    boost_expires_at: updated.boost_expires_at,
    boost_duration_days: updated.boost_duration_days,
    active: true,
  };
};

const boostWithCredits = async (vendorId, listingId, durationDays) => {
  const plan = await validatePlan(durationDays);
  const listing = await getOwnedListing(listingId, vendorId);

  await walletService.spendCredits(
    vendorId,
    plan.credits,
    `Boost listing ${durationDays}d`,
    'boost_listing',
    listingId
  );

  const boost = await applyBoost(listingId, durationDays);

  await notificationService.create(
    vendorId,
    'boost',
    'Listing boosted!',
    `'${listing.title}' is boosted for ${durationDays} days.`,
    {},
    `/listing/${listing.slug}`,
    'vendor'
  );

  return { listing_id: listingId, boost, payment_method: 'credits' };
};

const boostWithInr = async (vendorId, listingId, durationDays) => {
  const plan = await validatePlan(durationDays);
  await getOwnedListing(listingId, vendorId);
  const paymentService = require('./payment.service');
  const order = await paymentService.createPaymentOrder(
    vendorId,
    'listing_boost',
    plan.paise,
    `${listingId}:${durationDays}`
  );
  return {
    listing_id: listingId,
    payment: order,
    payment_method: 'inr',
    duration_days: durationDays,
  };
};

const activateBoostFromPayment = async (payment) => {
  const ref = payment.ref_id || '';
  if (!ref.includes(':')) {
    return null;
  }
  const parts = ref.split(':');
  const listingId = parts[0];
  const durationDays = parseInt(parts[1], 10);
  if (isNaN(durationDays)) {
    return null;
  }

  const boost = await applyBoost(listingId, durationDays);
  await notificationService.create(
    payment.user_id,
    'boost',
    'Boost activated',
    `Your listing is now boosted for ${durationDays} days.`,
    {},
    '/vendor/dashboard',
    'vendor'
  );
  return boost;
};

const listMyBoosted = async (vendorId) => {
  const nowIso = new Date().toISOString();
  const docs = await Listing.find({
    vendor_id: vendorId,
    is_deleted: { $ne: true },
    boost_expires_at: { $gt: nowIso },
  }).sort({ boost_expires_at: 1 });
  return docs.map(serializeListing);
};

const expireBoostsOnce = async () => {
  const nowIso = new Date().toISOString();
  
  // 1. Expire listing boosts
  const resListing = await Listing.updateMany(
    { boost_expires_at: { $lte: nowIso, $ne: null } },
    { $set: { boost_expires_at: null, boost_duration_days: null, updated_at: nowIso } }
  );

  // 2. Expire reel boosts
  const resReel = await Reel.updateMany(
    { boostExpiresAt: { $lte: new Date(), $ne: null }, isBoosted: true },
    { $set: { isBoosted: false, boostExpiresAt: null, boostDurationDays: null } }
  );

  return resListing.modifiedCount + resReel.modifiedCount;
};

const boostReelWithCredits = async (vendorId, reelId, durationDays) => {
  const days = parseInt(durationDays, 10);
  if (isNaN(days) || days <= 0) {
    throw ApiError.badRequest('Duration in days must be a positive number');
  }

  const reel = await Reel.findOne({ _id: reelId, isDeleted: { $ne: true } });
  if (!reel) {
    throw ApiError.notFound('Reel not found');
  }

  if (reel.creator.toString() !== vendorId.toString()) {
    throw ApiError.forbidden('Only the reel owner can boost');
  }

  // Fetch active credit rates to get cost per day
  let costPerDay = 10; // fallback default
  try {
    const rateSetting = await AppSettings.findOne({ key: 'credit_rates' }).lean();
    if (rateSetting && rateSetting.value && rateSetting.value.reelBoost1Day !== undefined) {
      costPerDay = Number(rateSetting.value.reelBoost1Day);
    }
  } catch (err) {
    logger.error('Failed to fetch credit rates for reel boosting:', err);
  }

  const totalCost = costPerDay * days;

  // Deduct credits from vendor's wallet
  await walletService.spendCredits(
    vendorId,
    totalCost,
    `Boost reel for ${days} days`,
    'boost_reel'
  );

  // Calculate new boost expiration date
  const now = new Date();
  let baseFrom = now;
  if (reel.boostExpiresAt && reel.boostExpiresAt > now) {
    baseFrom = reel.boostExpiresAt;
  }
  const newExpiry = new Date(baseFrom.getTime() + days * 24 * 60 * 60 * 1000);

  // Update reel properties
  await Reel.updateOne(
    { _id: reelId },
    {
      $set: {
        isBoosted: true,
        boostExpiresAt: newExpiry,
        boostDurationDays: days,
        boostActivatedAt: now,
        boostCost: totalCost,
      }
    }
  );

  // Trigger notification for vendor
  await notificationService.create(
    vendorId,
    'boost',
    'Reel boosted!',
    `Your reel is boosted for ${days} days.`,
    {},
    '/vendor/reels',
    'vendor'
  );

  return { reel_id: reelId, isBoosted: true, boostExpiresAt: newExpiry };
};

const expireBoostsLoop = async () => {
  try {
    const n = await expireBoostsOnce();
    if (n > 0) {
      logger.info(`Expired ${n} boost(s)`);
    }
  } catch (err) {
    logger.warn(`boost expire loop error: ${err.message}`);
  }
};

module.exports = {
  boostWithCredits,
  boostWithInr,
  activateBoostFromPayment,
  listMyBoosted,
  boostReelWithCredits,
  expireBoostsOnce,
  expireBoostsLoop,
};
